parsers/basic.c

```diff
@@ -83,6 +82,83 @@ static KeyWord basic_keywords[] = {
  *   FUNCTION DEFINITIONS
  */
 
+/* Match the name of a dim or const starting at pos. */
+static int extract_dim (char const *pos, vString * name, BasicKind kind)
+{
+	const char *old_pos = pos;
+	while (isspace (*pos))
+		pos++;
+
+	/* create tags only if there is some space between the keyword and the identifier */
+	if (old_pos == pos)
+		return 0;
+
+	vStringClear (name);
+
+	if (strncasecmp (pos, "shared", 6) == 0)
+		pos += 6; /* skip keyword "shared" */
+
+	while (isspace (*pos))
+		pos++;
+
+	/* capture "dim as String str" */
+	if (strncasecmp (pos, "as", 2) == 0)
+	{
+			pos += 2; /* skip keyword "as" */
+
+		while (isspace (*pos))
+			pos++;
+		while (!isspace (*pos)) /* skip next part which is a type */
+			pos++;
+		while (isspace (*pos))
+			pos++;
+		/* now we are at the name */
+	}
+	/* capture "dim as foo ptr bar" */
+	if (strncasecmp (pos, "ptr", 3) == 0 && isspace(*(pos+4)))
+	{
+		pos += 3; /* skip keyword "ptr" */
+		while (isspace (*pos))
+			pos++;
+	}
+	/*	capture "dim as string * 4096 chunk" */
+	if (strncmp (pos, "*", 1) == 0)
+	{
+		pos += 1; /* skip "*" */
+		while (isspace (*pos) || isdigit(*pos) || ispunct(*pos))
+			pos++;
+	}
+
+	for (; *pos && !isspace (*pos) && *pos != '(' && *pos != ',' && *pos != '='; pos++)
+		vStringPut (name, *pos);
+	makeSimpleTag (name, kind);
+
+	/* if the line contains a ',', we have multiple declarations */
+	while (*pos && strchr (pos, ','))
+	{
+		/* skip all we don't need(e.g. "..., new_array(5), " we skip "(5)") */
+		while (*pos != ',' && *pos != '\'')
+			pos++;
+
+		if (*pos == '\'')
+			return 0; /* break if we are in a comment */
+
+		while (isspace (*pos) || *pos == ',')
+			pos++;
+
+		if (*pos == '\'')
+			return 0; /* break if we are in a comment */
+
+		vStringClear (name);
+		for (; *pos && !isspace (*pos) && *pos != '(' && *pos != ',' && *pos != '='; pos++)
+			vStringPut (name, *pos);
+		makeSimpleTag (name, kind);
+	}
+
+	vStringDelete (name);
```

Comment:
Is it o.k. to delete `name` here?
I think this deletion should be at ...(B)

Reply:
After reading the last commit, putting `vStringDelete (name);` here makes sense.
However, I don't like these commit sequences. 

In this commit, I think you did two things: (1) fixing the memory management bug in extract_dim introduced in "basic: add more advanced extraction of dim-like variables", and (2) reorganizing the way to manage a vString storing "name".

(1) should be merged to  "basic: add more advanced extraction of dim-like variables".
So you can focus on (2) in this commit.

I think what I wrote here was a  kind of "Preaching to Buddha".